### PR TITLE
Add expandable list for leapp reports with details

### PR DESCRIPTION
**Pull request by fellipeh**
These PR add first version of expandable list for reports.

Need to add the remediations and the Fix remediation button.

Signed-off-by: Fellipe Henrique <fpedrosa@redhat.com>

### CODE REVIEW

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js [hunk 1]
```diff
@@ -1,16 +1,70 @@
...
+      <ListView.InfoItem key="k-host">
+        <Icon type="pf" name="cluster" /> {entry.hostname}
+      </ListView.InfoItem>,
+      <ListView.InfoItem key="k-severity">
+          <Icon
+              type="pf"
+              name="resources-full"
+              style={{ color: presentSeverity(entry.severity) }} /> {entry.severity}
+      </ListView.InfoItem>,
+      // hasRemediations(entry),
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js:23]: Can this be removed if not used?

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js [hunk 2]
```diff
@@ -1,16 +1,70 @@
...
+
+    ]}
+    // {additionalInfo(entry)}
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js:26]: Same as above.

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 3]
```diff
@@ -1,13 +1,39 @@
...
 import EmptyInfoItem from './EmptyInfoItem';
 import StringInfoItem from './StringInfoItem';
 
 export const itemIteratorId = (entry, ...rest) =>
   `${entry.id}-${entry.preupgradeReportId}-${rest.join('-')}`;
 
+export const hasRemediations = entry => {
+
+  if (entry.includes('detail')) {
+    let data = JSON.parse(entry.detail);
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:14]: Same as for tags, this is now serialized properly, so no need for `JSON.parse`.

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 4]
```diff
@@ -17,12 +43,28 @@ export const additionalInfo = entry => {
...
         return <StringInfoItem entry={entry} attr={attr} key={key} />;
       default:
         return '';
     }
   });
 };
+
+export const tagInfo = entry => {
+  
+  let data = JSON.parse(entry);
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:64]: I am getting: 

tags

Tags no longer have to be parsed as they are serialized properly: https://github.com/oamg/foreman_leapp/commit/0d734aee578c47b527e14614fc8f0679e26cb331

**Comment by xprazak2** [comment]
Nice progress. The current look:

expandable

Adding a bit of a space between rows in expandable content makes it look a bit spacious to my eyes, here is a version with 10px padding on rows:

with-padding


Also the description seems misaligned with the arrow. 

Note that js tests and lint are currently failing.

**Comment by adamruzicka** [comment]
Could you rebase so travis can check this?

**Comment by fellipeh** [comment]
> Could you rebase so travis can check this?

Rebased, finishing on next commit the changes @xprazak2 told..

**Comment by ares** [comment]
Nice! Not in this PR, however it was suggested on the call with UXD that we should try reusing the insights iconography for risk levels (and wherever applicable) for the consistency reasons. Just wanted to share that for the future work. Can't wait to see the UI in :-)

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 5]
```diff
@@ -1,13 +1,42 @@
...
+import { ListView, Badge, Icon } from 'patternfly-react';
+
 import EmptyInfoItem from './EmptyInfoItem';
 import StringInfoItem from './StringInfoItem';
 
 export const itemIteratorId = (entry, ...rest) =>
   `${entry.id}-${entry.preupgradeReportId}-${rest.join('-')}`;
 
+export const hasRemediations = entry => {
+  if (entry.includes('detail')) {
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:12]: `Uncaught TypeError: entry.includes is not a function`.

You can squash with the following `if` into one condition: `if (entry.detail && entry.detail.remediations)`

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 6]
```diff
@@ -1,13 +1,42 @@
...
+    // const data = JSON.parse(entry.detail);
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:13]: Can be removed as we will not be using it.

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js [hunk 7]
```diff
@@ -1,16 +1,64 @@
...
+    additionalInfo={[
+      <ListView.InfoItem key="k-host">
+        <Icon type="pf" name="cluster" /> {entry.hostname}
+      </ListView.InfoItem>,
+      <ListView.InfoItem key="k-severity">
+        <Icon
+          type="pf"
+          name="resources-full"
+          style={{ color: presentSeverity(entry.severity) }}
+        />{' '}
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js:22]: Is adding a whitespace nodes something patternfly recommends for creating a space between text and icon? Wouldn't it be better to use stylesheets for styling?

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 8]
```diff
@@ -17,12 +46,34 @@ export const additionalInfo = entry => {
...
+      case 'hostname':
         return <StringInfoItem entry={entry} attr={attr} key={key} />;
       default:
         return '';
     }
   });
 };
+
+export const tagInfo = entry => {
+  const data = JSON.parse(entry);
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:72]: Tags are now serialized, no need to parse them.

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js [hunk 9]
```diff
@@ -1,16 +1,64 @@
...
 const PreupgradeReportEntry = ({ entry }) => (
   <ListView.Item
     key={entry.id}
     className="listViewItem--listItemVariants"
     stacked
-    additionalInfo={additionalInfo(entry)}
-  />
+    description={entry.title}
+    additionalInfo={[
+      <ListView.InfoItem key="k-host">
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js:14]: `key` prop can be removed, I think it is needed only when using component inside iterator.

Comment (fellipeh) [reply webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js]: I'm still getting these error, and can't find where is it.. any clue?

```Warning: Each child in a list should have a unique "key" prop.

Check the render method of `PreupgradeReportEntry`. See https://fb.me/react-warning-keys for more information.
    in div (created by PreupgradeReportEntry)
    in PreupgradeReportEntry (created by PreupgradeReportsList)
    in div (created by ListView)
    in ListView (created by PreupgradeReportsList)
    in PreupgradeReportsList (created by PreupgradeReports)
    in LoadingState (created by PreupgradeReports)
    in PreupgradeReports (created by WrappedPreupgradeReports)
    in WrappedPreupgradeReports (created by I18nProviderWrapper(WrappedPreupgradeReports))
    in IntlProvider (created by I18nProviderWrapper(WrappedPreupgradeReports))
    in I18nProviderWrapper(WrappedPreupgradeReports) (created by StoreProvider(I18nProviderWrapper(WrappedPreupgradeReports)))
    in Provider (created by StoreProvider(I18nProviderWrapper(WrappedPreupgradeReports)))
    in StoreProvider(I18nProviderWrapper(WrappedPreupgradeReports)) (created by DataProvider(StoreProvider(I18nProviderWrapper(WrappedPreupgradeReports))))
    in DataProvider(StoreProvider(I18nProviderWrapper(WrappedPreupgradeReports)))```

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js [hunk 10]
```diff
@@ -1,16 +1,76 @@
...
+    ]}
+  >
+    <Grid fluid>
+      <Grid.Row>
+        <Grid.Col md={1}>
+          <strong>Title</strong>
+        </Grid.Col>
+        <Grid.Col md={8}>{entry.title}</Grid.Col>
+      </Grid.Row>
+      <Grid.Row style={{ marginTop: 10 }}>
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js:42]: Applying the desired margin via (s)css file would make this more DRY.

**Comment by xprazak2** [comment]
It looks pretty nice:

new-look

Margin is applied to empty rows, which results in rows with content having more space between them:

empty-row

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 11]
```diff
@@ -1,13 +1,28 @@
...
+    case 'medium':
+      return '#f0ab00';
+    case 'high':
+      return '#cc0000';
+    default:
+      return '#39a5dc';
+  }
+};
+
 export const additionalInfo = entry => {
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:24]: I think this is no longer used in `PreupgradeReportEntry` and can be removed.

File: webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js (hunk 9 above)
Comment (xprazak2) [reply webpack/components/PreupgradeReportsList/components/PreupgradeReportEntry.js]: My guess would be `getExternals` method as the returned markup items created in iterator do not have `key` property.

**Comment by xprazak2** [comment]
Travis is currently failing, it would be good to have tests for the newly added code.

File: webpack/components/PreupgradeReportsList/components/__tests__/EmptyInfoItem.test.js [hunk 12]
```diff
@@ -1,13 +1,14 @@
...
+// const fixtures = {
+//   'should render': {
+//     entry: { title: 'no title', severity: 'low', id: 5, preupgradeReportId: 7 },
+//     attr: 'title',
+//   },
+// };
 
 describe('EmptyInfoItem', () =>
-  testComponentSnapshotsWithFixtures(EmptyInfoItem, fixtures));
+  testComponentSnapshotsWithFixtures(EmptyInfoItem, reportEntry));
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/__tests__/EmptyInfoItem.test.js:14]: The props sturucture for `EmptyInfoItem` has not changed, so you can keep the fixture as is. You can also replace the current `entry` in `fixtures` with the `reportEntry` if you need to.

File: webpack/components/PreupgradeReportsList/components/helpers.js [hunk 13]
```diff
@@ -1,28 +1,157 @@
...
+      }
+      if (attr.type === 'command') {
+        return (
+          <Grid.Row className="top-padded" key={key}>
+            <Grid.Col md={1}>
+              <strong>Command</strong>
+            </Grid.Col>
+            <Grid.Col md={8}>
+              <code>{attr.context.join(' ')}</code>{' '}
+              <Button bsSize="xsmall">Run Remediation</Button>
```
Comment (xprazak2) [inline webpack/components/PreupgradeReportsList/components/helpers.js:147]: No need for the button, we will not be running each remediation separately. #24 and #32 will take care of it.

**Comment by xprazak2** [comment]
Nice progress, almost there.

**Comment by xprazak2** [comment]
I think `StringInfoItem` component is no longer used, if that is really the case, I think we should remove it.

**Comment by xprazak2** [comment]
Looks good, seems like some snapshots need an update.

**Comment by xprazak2** [comment]
Thanks!

**Pull request merged**